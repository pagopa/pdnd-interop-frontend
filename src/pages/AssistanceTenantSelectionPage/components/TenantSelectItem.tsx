import React from 'react'
import {
  Avatar,
  ListItem,
  ListItemAvatar,
  ListItemButton,
  type ListItemProps,
  ListItemText,
} from '@mui/material'
import AccountBalanceIcon from '@mui/icons-material/AccountBalance'
import type { CompactTenant } from '@/api/api.generatedTypes'
import { useTranslation } from 'react-i18next'

type TenantSelectItemProps = {
  component?: React.ElementType
  tenant: CompactTenant
  onClick?: VoidFunction
} & ListItemProps

export const TenantSelectItem: React.FC<TenantSelectItemProps> = ({
  onClick,
  component,
  tenant,
  ...listItemProps
}) => {
  const { t } = useTranslation('common')

  const Wrapper = onClick
    ? ({ children }: { children: React.ReactNode }) => (
        <ListItemButton onClick={onClick} sx={{ borderRadius: 1 }}>
          {children}
        </ListItemButton>
      )
    : React.Fragment

  return (
    <ListItem
      /**
       * For some reason the `component` prop is not recognized by TypeScript
       * even if it's a valid prop of ListItem component.
       * See https://github.com/mui/material-ui/blob/master/packages/mui-material/src/ListItem/ListItem.js#L180
       */
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      component={component}
      disablePadding
      {...listItemProps}
    >
      <Wrapper>
        <ListItemAvatar>
          <Avatar sx={{ width: 48, height: 48, bgcolor: 'white', border: 1 }} src={tenant?.logoUrl}>
            <AccountBalanceIcon sx={{ color: '#bdbdbd' }} />
          </Avatar>
        </ListItemAvatar>
        <ListItemText primary={tenant.name} secondary={t('userProductRole.admin')} />
      </Wrapper>
    </ListItem>
  )
}
