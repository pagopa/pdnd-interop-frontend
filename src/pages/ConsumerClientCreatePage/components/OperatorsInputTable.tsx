import React from 'react'
import { Table, TableRow } from '@/components/shared/Table'
import { Button, IconButton, Stack } from '@mui/material'
import { useFormContext } from 'react-hook-form'
import { useTranslation } from 'react-i18next'
import type { CreateClientFormValues } from '../ConsumerClientCreate.page'
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline'
import { useDialog } from '@/stores'
import type { SelfCareUser } from '@/types/party.types'

const OperatorsInputTable: React.FC = () => {
  const { t } = useTranslation('client')
  const { t: tCommon } = useTranslation('common')
  const { openDialog } = useDialog()

  const { watch, setValue } = useFormContext<CreateClientFormValues>()
  const operators = watch('operators')

  const headLabels = [tCommon('table.headData.userName'), '']

  const handleRemoveOperator = (operatorId: string) => {
    const newOperators = operators.filter((o) => o.id !== operatorId)
    setValue('operators', newOperators)
  }

  const handleAddOperator = (newOperators: Array<SelfCareUser>) => {
    setValue('operators', [...operators, ...newOperators])
  }

  const handleAddOperatorDialog = () => {
    openDialog({
      type: 'addSecurityOperator',
      excludeOperatorsIdsList: operators.map(({ id }) => id),
      onSubmit: handleAddOperator,
    })
  }

  return (
    <>
      <Table
        isEmpty={operators.length === 0}
        headLabels={headLabels}
        noDataLabel={t('create.operatorsTable.noDataLabel')}
      >
        {operators.map((operator) => (
          <TableRow
            key={operator.id}
            cellData={[{ label: `${operator.name} ${operator.familyName}` }]}
          >
            <IconButton onClick={handleRemoveOperator.bind(null, operator.id)}>
              <DeleteOutlineIcon color="error" fontSize="small" />
            </IconButton>
          </TableRow>
        ))}
      </Table>
      <Stack direction="row" sx={{ my: 2 }}>
        <Button type="button" variant="contained" size="small" onClick={handleAddOperatorDialog}>
          {tCommon('addBtn')}
        </Button>
      </Stack>
    </>
  )
}

export default OperatorsInputTable
