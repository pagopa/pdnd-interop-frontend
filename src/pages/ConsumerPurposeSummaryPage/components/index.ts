export * from './ConsumerPurposeSummaryGeneralInformationAccordion'
export * from './ConsumerPurposeSummaryRiskAnalysisAccordion'
