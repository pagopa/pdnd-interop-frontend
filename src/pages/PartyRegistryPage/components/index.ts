export * from './PartyAttributesSection'
export * from './PartyContactsSection'
