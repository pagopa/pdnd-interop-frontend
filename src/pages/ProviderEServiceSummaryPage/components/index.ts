export * from './ProviderEServiceGeneralInfoSummary'
export * from './ProviderEServiceVersionInfoSummary'
export * from './ProviderEServiceDocumentationSummary'
export * from './ProviderEServiceRiskAnalysisSummary'
