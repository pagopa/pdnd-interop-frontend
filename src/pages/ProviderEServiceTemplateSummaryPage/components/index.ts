export * from './ProviderEServiceTemplateGeneralInfoSummary'
export * from './ProviderEServiceTemplateVersionInfoSummary'
export * from './ProviderEServiceTemplateAttributeVersionSummary'
export * from './ProviderEServiceTemplateDocumentationSummary'
export * from './ProviderEServiceTemplateRiskAnalysisSummary'
