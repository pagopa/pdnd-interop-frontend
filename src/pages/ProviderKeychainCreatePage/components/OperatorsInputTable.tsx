import React from 'react'
import { Button, Stack } from '@mui/material'
import { useFormContext } from 'react-hook-form'
import { useTranslation } from 'react-i18next'
import type { CreateKeychainFormValues } from '../ProviderKeychainCreate.page'
import { Table, TableRow } from '@pagopa/interop-fe-commons'
import PlusOneIcon from '@mui/icons-material/PlusOne'
import { useDrawerState } from '@/hooks/useDrawerState'
import type { Users } from '@/api/api.generatedTypes'
import { useCurrentRoute } from '@/router'
import { AddOperatorsDrawer } from '@/components/shared/AddOperatorsDrawer'

const OperatorsInputTable: React.FC = () => {
  const { t } = useTranslation('client')
  const { t: tCommon } = useTranslation('common')

  const {
    isOpen: isAddOperatorDrawerOpen,
    openDrawer: openAddOperatorDrawer,
    closeDrawer: closeAddOperatorDrawer,
  } = useDrawerState()

  const { watch, setValue } = useFormContext<CreateKeychainFormValues>()
  const operators = watch('operators')

  const headLabels = [tCommon('table.headData.userName'), '']

  const handleRemoveOperator = (operatorId: string) => {
    const newOperators = operators.filter((o) => o.userId !== operatorId)
    setValue('operators', newOperators)
  }

  const handleAddOperator = (newOperators: Users) => {
    setValue('operators', [...operators, ...newOperators])
  }

  const handleOpenAddOperatorDrawer = () => {
    openAddOperatorDrawer()
  }

  const { routeKey } = useCurrentRoute()

  return (
    <>
      <Table
        isEmpty={operators.length === 0}
        headLabels={headLabels}
        noDataLabel={t('create.operatorsTable.noDataLabel')}
      >
        {operators.map((operator) => (
          <TableRow key={operator.userId} cellData={[`${operator.name} ${operator.familyName}`]}>
            <Button
              onClick={handleRemoveOperator.bind(null, operator.userId)}
              variant="naked"
              color="error"
            >
              {tCommon('actions.delete')}
            </Button>
          </TableRow>
        ))}
      </Table>
      <Stack direction="row" sx={{ my: 2 }}>
        <Button
          type="button"
          variant="contained"
          size="small"
          onClick={handleOpenAddOperatorDrawer}
          startIcon={<PlusOneIcon />}
        >
          {tCommon('addBtn')}
        </Button>
      </Stack>
      <AddOperatorsDrawer
        isOpen={isAddOperatorDrawerOpen}
        onClose={closeAddOperatorDrawer}
        excludeOperatorsIdsList={operators.map(({ userId }) => userId)}
        onSubmit={handleAddOperator}
        operatorKind="keychain"
      />
    </>
  )
}

export default OperatorsInputTable
