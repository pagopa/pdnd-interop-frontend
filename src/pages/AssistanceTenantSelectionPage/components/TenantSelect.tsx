import React from 'react'
import {
  IconButton,
  InputAdornment,
  List,
  Paper,
  Skeleton,
  TextField,
  Typography,
} from '@mui/material'
import CloseIcon from '@mui/icons-material/Close'
import { TenantSelectItem } from './TenantSelectItem'
import SearchIcon from '@mui/icons-material/Search'
import { useTranslation } from 'react-i18next'
import { PartyQueries } from '@/api/party'
import { useAutocompleteTextInput } from '@pagopa/interop-fe-commons'
import type { CompactTenant } from '@/api/api.generatedTypes'

type TenantSelectProps = {
  selectedTenant: CompactTenant | null
  onSelectTenant: (value: CompactTenant | null) => void
}

export const TenantSelect: React.FC<TenantSelectProps> = ({ selectedTenant, onSelectTenant }) => {
  const { t } = useTranslation('assistance', { keyPrefix: 'tenantSelection' })

  const [query, setQuery] = useAutocompleteTextInput()
  const { data: tenants } = PartyQueries.useGetTenants(
    {
      name: query,
      limit: 50,
    },
    {
      keepPreviousData: true,
    }
  )

  const tenantsOptions = tenants?.results ?? []

  if (selectedTenant) {
    return (
      <TenantSelectWrapper>
        <TenantSelectItem
          component="div"
          tenant={selectedTenant}
          secondaryAction={
            <IconButton
              onClick={() => onSelectTenant(null)}
              edge="end"
              aria-label={t('deselectButtonLabel')}
            >
              <CloseIcon />
            </IconButton>
          }
        />
      </TenantSelectWrapper>
    )
  }

  return (
    <TenantSelectWrapper>
      <TextField
        fullWidth
        size="small"
        label={t('textFieldLabel')}
        autoFocus
        onChange={(e) => setQuery(e.target.value)}
        InputProps={{
          startAdornment: (
            <InputAdornment position="start">
              <SearchIcon />
            </InputAdornment>
          ),
        }}
      />
      <List
        sx={{
          mt: 2,
          maxHeight: 220,
          overflowY: 'auto',
          scrollbarWidth: 'thin',
          '&::-webkit-scrollbar': {
            width: '4px',
          },
          '&::-webkit-scrollbar-track': {
            boxShadow: 'rgb(230, 233, 242) 10px 10px inset',
          },
          '&::-webkit-scrollbar-thumb': {
            backgroundColor: 'rgb(0, 115, 230)',
            borderRadius: '16px',
          },
          '&::-webkit-scrollbar-thumb:hover': {
            background: '#555',
          },
        }}
      >
        {tenantsOptions.map((tenantOption) => (
          <TenantSelectItem
            key={tenantOption.id}
            tenant={tenantOption}
            onClick={() => onSelectTenant(tenantOption)}
          />
        ))}
        {tenantsOptions.length === 0 && (
          <Typography fontWeight={700}>{t('noResultsLabel')}</Typography>
        )}
      </List>
    </TenantSelectWrapper>
  )
}

export const TenantSelectSkeleton: React.FC = () => {
  return (
    <Skeleton
      variant="rectangular"
      sx={{ maxWidth: 480, height: 343, width: '100%', borderRadius: 4 }}
    />
  )
}

const TenantSelectWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <Paper elevation={8} sx={{ maxWidth: 480, p: 4, width: '100%', borderRadius: 4 }}>
      {children}
    </Paper>
  )
}
