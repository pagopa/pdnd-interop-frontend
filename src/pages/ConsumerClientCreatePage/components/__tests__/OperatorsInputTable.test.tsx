import { BACKEND_FOR_FRONTEND_URL } from '@/config/env'
import { mockUseJwt, queryClientMock, setupQueryServer } from '@/utils/testing.utils'
import { render } from '@testing-library/react'
import { createMockSelfCareUser } from '__mocks__/data/user.mocks'
import OperatorsInputTable from '../OperatorsInputTable'
import React from 'react'
import type { CreateClientFormValues } from '../../ConsumerClientCreate.page'
import { FormProvider, useForm } from 'react-hook-form'
import type { RelationshipInfo } from '@/api/api.generatedTypes'
import userEvent from '@testing-library/user-event'
import { Dialog } from '@/components/dialogs'
import { QueryClientProvider } from '@tanstack/react-query'

const queryServer = setupQueryServer([
  {
    url: `${BACKEND_FOR_FRONTEND_URL}/tenants/:tenantId/relationships`,
    result: [
      createMockSelfCareUser({ id: '1', name: 'operator 1' }),
      createMockSelfCareUser({ id: '2', name: 'operator 2' }),
    ],
  },
])

beforeAll(() => queryServer.listen())
afterEach(() => queryServer.resetHandlers())
afterAll(() => queryServer.close())

const renderOperatorsInputTable = (operators: Array<RelationshipInfo> = []) => {
  const Wrapper = ({ children }: { children: React.ReactNode }) => {
    const useFormMethods = useForm<CreateClientFormValues>({
      defaultValues: { name: '', description: '', operators },
    })

    return (
      <QueryClientProvider client={queryClientMock}>
        <Dialog />
        <FormProvider {...useFormMethods}>{children}</FormProvider>
      </QueryClientProvider>
    )
  }

  return render(<OperatorsInputTable />, { wrapper: Wrapper })
}

describe('OperatorsInputTable', () => {
  it('should match the snapshot with no selected operators', async () => {
    const screen = renderOperatorsInputTable()
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should match the snapshot with selected operators', async () => {
    const screen = renderOperatorsInputTable([createMockSelfCareUser()])
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should correctly remove operators', async () => {
    const screen = renderOperatorsInputTable([createMockSelfCareUser()])
    const user = userEvent.setup()
    const removeButton = screen.getByRole('button', { name: 'actions.delete' })

    const operatorCell = screen.getByRole('cell', { name: 'Mario Rossi' })
    expect(operatorCell).toBeInTheDocument()
    await user.click(removeButton)
    expect(operatorCell).not.toBeInTheDocument()
  })

  it('should correctly add operators', async () => {
    mockUseJwt()
    const screen = renderOperatorsInputTable()
    const user = userEvent.setup()
    const addButton = screen.getByRole('button', { name: 'addBtn' })

    await user.click(addButton)
    const operatorAutocomplete = screen.getByLabelText('content.autocompleteLabel')
    await user.click(operatorAutocomplete)
    const option = screen.getByRole('option', { name: 'operator 1 Rossi' })
    await user.click(option)

    const confirmButton = screen.getByRole('button', { name: 'actions.confirmLabel' })
    await user.click(confirmButton)

    expect(screen.getByRole('cell', { name: 'operator 1 Rossi' })).toBeInTheDocument()
  })
})
