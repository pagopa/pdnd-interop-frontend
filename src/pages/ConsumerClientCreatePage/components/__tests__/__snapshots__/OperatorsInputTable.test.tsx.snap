// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OperatorsInputTable > should match the snapshot with no selected operators 1`] = `
<body>
  <div>
    <div
      class="MuiTableContainer-root css-u77c0j-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-whv87p-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              table.headData.userName
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-1nq9kkf-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-d9l1z9-MuiTableCell-root"
              colspan="2"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-1fuz9qr-MuiPaper-root-MuiAlert-root"
                role="alert"
              >
                <div
                  class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="InfoOutlinedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
                >
                  create.operatorsTable.noDataLabel
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="css-r4oyyn-MuiStack-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        addBtn
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`OperatorsInputTable > should match the snapshot with selected operators 1`] = `
<body>
  <div>
    <div
      class="MuiTableContainer-root css-u77c0j-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-whv87p-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              table.headData.userName
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-1nq9kkf-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1p72hwd-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1rpd3i8-MuiTableCell-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
              >
                Mario Rossi
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-2u9kxv-MuiTableCell-root"
            >
              <button
                aria-label="actions.delete"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeSmall css-1efkf0c-MuiSvgIcon-root"
                  data-testid="DeleteOutlineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="css-r4oyyn-MuiStack-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        addBtn
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;
